### PR TITLE
fix: Make the default backend RFC 8259 compliant

### PR DESCRIPTION
**Pull request by keelerm84**
BEGIN_COMMIT_OVERRIDE
chore: Make the default backend RFC 8259 compliant
END_COMMIT_OVERRIDE

The default (non-easyjson) tokenizer accepted several inputs that RFC 8259
forbids and mangled one valid input. Since the default implementation is the
source of truth for this library, align its accept/reject decisions with
encoding/json's json.Valid.

Reader (jreader/token_reader_default.go):
- Whitespace: allow only space, tab, LF, and CR between tokens instead of the
  broader unicode.IsSpace (which accepted vertical tab, form feed, NEL, NBSP).
- Numbers: reject leading zeros, a bare minus sign, and a trailing decimal
  point by walking the RFC number grammar.
- Strings: reject unescaped control characters (below 0x20).
- Strings: combine UTF-16 surrogate pairs into a single code point; lone or
  invalid surrogates decode to the replacement character, matching
  encoding/json.

Writer (jwriter/token_writer_default.go):
- Float64 returns an error for NaN and infinities rather than emitting invalid
  JSON.

Tests are default-only because the easyjson backend does not make all of the
same decisions.

---

> [!NOTE]
> **Medium Risk**
> Stricter parsing and different edge-case decoding may reject or change values for inputs the old tokenizer accepted; scope is core JSON I/O but behavior is intentionally aligned with encoding/json.
> 
> **Overview**
> Aligns the **default** (non-easyjson) JSON tokenizer and writer with **RFC 8259** and **`encoding/json`** accept/reject and decode behavior.
> 
> **Reader:** Token whitespace is limited to space, tab, LF, and CR (no longer all `unicode.IsSpace`). Number parsing follows the RFC grammar (e.g. rejects leading zeros, bare `-`, trailing `.`) and **int64 overflow** falls back to float parsing instead of wrapping. Strings reject unescaped controls; invalid UTF-8 and lone/malformed `\u` surrogates match stdlib (replacement char); valid surrogate pairs combine into one code point.
> 
> **Writer:** `Float64` returns an error for **NaN** and **±Inf** instead of emitting invalid JSON.
> 
> Adds **default-only** RFC compliance tests (easyjson backend unchanged).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c14f91633c99cf4866d318f481b3ba1f8ba453e3. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: jreader/token_reader_rfc8259_default_test.go [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+			assert.Equal(t, tc.want, got)
+		})
+	}
+
+	// The same pair encoded as literal UTF-8 bytes must also decode intact.
+	r := NewReader([]byte(`"a𝄞b"`))
+	got := r.String()
+	require.NoError(t, r.Error())
+	require.NoError(t, r.RequireEOF())
+	assert.Equal(t, "a\U0001D11Eb", got)
```
Comment (aaron-zeisler) [inline jreader/token_reader_rfc8259_default_test.go:178]: Minor: I think these 5 lines are duplicate code - it looks like the exact same thing as the 2nd test case.

**Review by aaron-zeisler — APPROVED** [review]
Approved with a minor recommendation 👍  To me, this code feels finicky, but I understand its purpose, and I'm not sure how to improve it generally.

**Review by kinyoklion — APPROVED** [review]
Potential concern about non-breaking spaces, but it is JSON compliant.

**Comment by kinyoklion** [comment]
> Potential concern about non-breaking spaces, but it is JSON compliant.

Looks like at least some other SDKs also reject, so we should be good.
